all_seaing_autonomy/all_seaing_autonomy/roboboat/WaypointFinder.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
+import rclpy
+from rclpy.node import Node
+from all_seaing_interfaces.msg import ObstacleMap, Obstacle
+from geometry_msgs.msg import Point, Pose, PoseArray, Vector3, Quaternion
+from all_seaing_interfaces.msg import BuoyPair, BuoyPairArray, Waypoint, WaypointArray
+from tf2_msgs.msg import TFMessage
+from nav_msgs.msg import Odometry
+from visualization_msgs.msg import Marker, MarkerArray
+from std_msgs.msg import Header, ColorRGBA
+import math
+
+class WaypointFinder(Node):
+    def __init__(self):
+        super().__init__("waypoint_finder")
+        self.map_sub = self.create_subscription(
+            ObstacleMap, "/obstacle_map/labeled", self.map_cb, 10
+        )
+        self.odometry_sub = self.create_subscription(Odometry, "/odometry/filtered", self.odometry_cb, 10)
+        self.buoy_pair_pub = self.create_publisher(BuoyPairArray, "buoy_pairs", 10)
+        self.waypoint_pub = self.create_publisher(WaypointArray, "waypoints", 10)
+        self.waypoint_marker_pub = self.create_publisher(MarkerArray, "waypoint_markers", 10)
+
+        self.robot_pos = (0,0)
+        self.safe_margin = 0
+
+        self.first_map = True
+
+    def norm_squared(self, vec, ref=(0, 0)):
+        return vec[0] ** 2 + vec[1] ** 2
+
+    def get_closest_to(self, source, points):
+        return min(points, key=lambda point: self.norm_squared(source, point))
+
+    def midpoint(self, vec1, vec2):
+        return ((vec1[0] + vec2[0]) / 2, (vec1[1] + vec2[1]) / 2)
+    
+    def midpoint_pair(self, pair):
+        return self.midpoint(pair[0], pair[1])
+    
+    def split_buoys(self, obstacles):
```

Comment:
rather than hardcoding green = 2 and red = 1, can you read in the labels for the colors from this [config file](https://github.com/ArcturusNavigation/all_seaing_vehicle/blob/main/all_seaing_bringup/config/perception/color_label_mappings.yaml) (passed in as a parameter)? for an example of how to do this, you can take a look at the constructor of the color segmentation node + how color segmentation is called in the launch file.